src/internal_telemetry/allocations/allocator/tracing.rs

```diff
@@ -0,0 +1,90 @@
+use std::{any::TypeId, marker::PhantomData, ptr::addr_of};
+
+use tracing::{Dispatch, Id, Subscriber};
+use tracing_subscriber::{layer::Context, registry::LookupSpan, Layer};
+
+use super::AllocationGroupToken;
+
+pub(crate) struct WithAllocationGroup {
+    pub with_allocation_group: fn(&Dispatch, &Id, AllocationGroupToken),
+}
+
+/// [`AllocationLayer`] is a [`tracing_subscriber::Layer`] that handles entering and exiting an allocation
+/// group as the span it is attached to is itself entered and exited.
+///
+/// More information on using this layer can be found in the examples, or directly in the
+/// `tracing_subscriber` docs, found [here][tracing_subscriber::layer].
+#[cfg_attr(docsrs, doc(cfg(feature = "tracing-compat")))]
+pub struct AllocationLayer<S> {
+    ctx: WithAllocationGroup,
+    _subscriber: PhantomData<fn(S)>,
+}
+
+impl<S> AllocationLayer<S>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+{
+    /// Creates a new [`AllocationLayer`].
+    #[must_use]
+    pub fn new() -> Self {
+        let ctx = WithAllocationGroup {
+            with_allocation_group: Self::with_allocation_group,
```

Comment:
It might be worth leaving a comment here that this method reference is specialized on `S` and so isn't just a single function reference but one for each `S` used in the program.